content/courses/ada-idioms/chapters/inheritance_idioms.rst

```diff
@@ -0,0 +1,577 @@
+private type that includes the fact of the derivation from the parent type.
+Otherwise, clients have no view of the parent type. Whether or not the child is
+visibly derived, the representation is not compile-time visible to clients, as
+for any private type. For example, type :ada:`Foo` is not visibly derived in the
+following:
+
+.. code-block:: ada
+
+    package P is
+       type Foo is tagged private;  --  visibly tagged but not visibly derived
+       -- ...
+    end P;
+
+To be visibly derived, we declare the child type as a private type using a
+*private extension*. A private extension is like a type extension, in that it
+introduces the child type name and the parent type. But like any private type
+declaration, it does not specify the type's representation. This is the first of
+the two steps for declaring a private type; hence it appears in the package
+visible part. For example:
+
+.. code-block:: ada
+
+    with ...
+    package P is
+       type Child is new Parent with private;  -- visibly derived from Parent
+    private
+       type Child is new Parent with record ... end record;
+    end P;
+
+The representation additions are not expressed until the private type's
+completion in the package private part, using a type extension. The steps are
+the same two for any private type: a declaration in the package visible part,
+with a completion in the package private part. The difference is the client
+visibility to the parent type.
+
+
+Solution
+--------
+
+There are two *solutions* in this entry, one for each of the two inheritance
+idioms under discussion. First, we will specify our building block choices,
+then we will show the two idiom expressions in separate subsections.
+
+- We will use tagged types for the sake of the full OOP capabilities. That is
+  the most common choice when inheritance is involved. The static form of
+  inheritance has cases in which it is useful, however those cases are very
+  narrow in applicability.
+
+- We will assume that both the parent type and the child type are both private
+  types, i.e., abstract data types, because that is the best practice. See the
+  :ref:`Abstract Data Type idiom <Ada_Idioms_Abstract_Data_Types>` for
+  justification and details.
+
+- For the most general capabilities, we will assume that the parent type is
+  visibly tagged.
+
+- We are going to declare the child type in a distinct, dedicated package,
+  following the :ref:`ADT idiom <Ada_Idioms_Abstract_Data_Types>`. This package
+  may or may not be a hierarchical child of the parent package. The solution
+  approach does not require a child package's special compile-time visibility,
+  although a child package is often necessary for the sake of that visibility.
+
+- Whether the child type is visibly derived will vary with the
+  :ref:`inheritance idiom <Ada_Idioms_Implementation_Inheritance>` solution.
+
+To avoid unnecessary code duplication, the two idiom solutions use the same
```

Comment:
"duplication, the two idiom solutions" should be "duplication, examples for the two idiom solutions"